### PR TITLE
Making the test document current

### PR DESCRIPTION
**Pull request by mbruzek**
Updating and writing testing

### CODE REVIEW

**Comment by evilnick** [comment]
Hello! I am unsure why this is a reference document. It doesn't look like reference material to me...

**Comment by mbruzek** [comment]
Good feedback @evilnick.  The idea was to write a testing document that was more recent than the one that currently exists.  I can change it to `developer-testing.md` or something similar based on your feedback.

**Comment by chuckbutler** [comment]
Category :cop:

File: src/en/developer-getting-started.md [hunk 1]
```diff
@@ -133,23 +133,17 @@ juju expose vanilla
...
-for example) which installs Juju if not already installed and any other packages
-required for testing.
-
-The other files in the test directory can now be written in Python and use the
-Amulet library.  The tests should deploy the charm, change configuration, relate
-to other charms where appropriate.  The test should ensure the service is
-running, operational, received the configuration from the test and was
-successfully related to the other charms.
+For more information about writing tests please refer to the
+[testing reference document](./reference-testing.md).
```
Comment (chuckbutler) [inline src/en/developer-getting-started.md]: Developer Charm Testing Guide maybe?

Comment (chuckbutler) [inline src/en/developer-getting-started.md]: category :cop:

File: src/en/developer-testing.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  intervention to get a successful test result.
+
+If a tool is needed to perform a test and is not available in the Ubuntu
+archive, it can also be included in the `tests/` directory, as long as the file
+which contains it is not executable. Note that build tools cannot be assumed to
+be available on the testing system.
+
+## Test automation
+
+The test runner will be run automatically so all charm tests must be self
```
Comment (chuckbutler) [inline src/en/developer-testing.md]: This statement doesn't seem so obvious. What about this is automatic? Whats running it?

This seems more like a lead in to talking about our CI infrastructure thats part of becoming a ~recommended charm, and test automation best practices...

File: src/en/developer-testing.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+contained, meaning tests must install or package the files required to test
+the charm.  The test runner will find and execute each test within that
+directory and produce a report.
+
+If tests exit with services still in the environment, the test runner may clean
+them up, whether by destroying the environment or destroying the services
+explicitly, and the machines may be terminated as well. Any artifacts needed
+from the test machines should be retrieved and displayed before the test exits.
+
```
Comment (chuckbutler) [inline src/en/developer-testing.md]: +1 to this ^

File: src/en/developer-testing.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+- 1: Failed test
+- 100: Test is skipped because of incomplete environment
+
+## charm proof
+
+The `charm-tools` package contains a static charm analysis tool called
+`charm proof`.  This tool checks the charm structure and gives Informational,
+Warning, and Error messages on potential issues with the charm structure. All
+charms should pass `charm proof` with Information messages only.
+Warning or Error messages indicate a problem that should be fixed in the charm.
```
Comment (chuckbutler) [inline src/en/developer-testing.md]: Maybe wrong place/time - but i stlll think anything emitting Warn should be moved to a soon to be implemented policy change, all the current W's be moved to E's as they are blockers to acceptance.

File: src/en/developer-testing.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+## Amulet
+
+While you can write tests in Bash or other languages, the
+[Amulet library](./tools-amulet.html) makes it easy to write tests in Python and is
+recommended.
+
```
Comment (chuckbutler) [inline src/en/developer-testing.md]: Amulet in itself deserves a page. Maybe link to a TODO page for now?

File: src/en/developer-testing.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+## BundleTester
+
+The charm test runner is a tool called
+[`bundletester`](https://github.com/juju-solutions/bundletester). The
+`bundletester` tool is used to find, fetch, and run tests on charms and
+[bundles](./charms-bundles.html).
+
```
Comment (chuckbutler) [inline src/en/developer-testing.md]: Can we make bundletester its own tool page under that reference section, and copy in the bundletester docs/readme?

File: src/en/developer-testing.md (hunk 4 above)
Comment (mbruzek) [inline src/en/developer-testing.md]: Would it help if I linked to the best practices or charm store policy documents?

File: src/en/developer-testing.md (hunk 5 above)
Comment (mbruzek) [inline src/en/developer-testing.md]: Amulet has a page, if that needs to be updated we can add a card to the board.

File: src/en/developer-testing.md (hunk 6 above)
Comment (mbruzek) [inline src/en/developer-testing.md]: Good idea!  We need someone else to write the bundletester document.

File: src/en/developer-testing.md (hunk 2 above)
Comment (mbruzek) [inline src/en/developer-testing.md]: This was not clear to IBM, they want to provide username and password on some tests, or manually run tests.

**Comment by chuckbutler** [comment]
Approval :cop: :+1:

**Pull request merged**